File: wltwallet/probe.go
// Wallet:probeActivity — takes a mnemonic-backed wallet, derives
// the canonical addresses for every supported chain (+ standard
// BIP variants where applicable), probes each RPC for on-chain
// activity, and returns the candidate list to the caller.
//
// Read-only. No Account rows created, no state changed. Callers
// use the returned candidates to decide:
//   - which Accounts to create on the mnemonic wallet itself
//     (for keep-as-mnemonic usage), or
//   - which chains to migrate to MPC via Wallet:promote.

package wltwallet

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"strings"
	"sync"
	"time"

	"github.com/KarpelesLab/apirouter"
	"github.com/KarpelesLab/base58"
	"github.com/KarpelesLab/libwallet/wltintf"
	"github.com/KarpelesLab/libwallet/wltnet"
	"github.com/KarpelesLab/libwallet/wltsign"
	"github.com/KarpelesLab/outscript"
	"github.com/KarpelesLab/secp256k1"
	"github.com/tyler-smith/go-bip39"
)

// probeCandidate describes one (chain, BIP variant) combination the
// probe will try. The address is computed from the curve-appropriate
// pubkey of derivePrivkeyFromSeed(seed, curve, Path); the probeFn
// then hits the chain's RPC to detect activity.
type probeCandidate struct {
	Network        string // human-readable chain tag: "bitcoin" | "litecoin" | "monacoin" | "bitcoin-cash" | "dogecoin" | "ethereum" | "solana"
	Variant        string // short label for UI: "P2PKH" | "P2WPKH (BIP84)" | "standard" | "phantom" | "sollet"
	Curve          string // "secp256k1" | "ed25519"
	DerivationPath string // BIP32 path, e.g. "m/84'/0'/0'/0/0" (empty string means curve-specific default)
	outputScript   string // outscript.New(pub).Out(...) — "p2wpkh" / "p2pkh" / "eth" (ignored for solana)
	addressChain   string // outscript .Address(...) chain tag
	networkType    string // wltnet.Network.Type for the probe call
	networkChainId string // wltnet.Network.ChainId for the probe call
}

// ProbeResult is one row in the Wallet:probeActivity response.
type ProbeResult struct {
	Network        string `json:"network"`                  // "bitcoin" | "ethereum" | ...
	Variant        string `json:"variant"`                  // "P2WPKH (BIP84)" etc.
	Curve          string `json:"curve"`                    // "secp256k1" | "ed25519"
	DerivationPath string `json:"derivationPath"`           // path that produced the address
	Address        string `json:"address"`                  // derived address
	Pubkey         string `json:"pubkey"`                   // base64url-encoded (compressed for secp, raw for ed25519)
	HasActivity    bool   `json:"hasActivity"`              // true if balance > 0 or any tx history
	Balance        string `json:"balance,omitempty"`        // raw balance units (wei, satoshi, lamports). Empty on probe error.
	Error          string `json:"error,omitempty"`          // per-candidate probe error (stringified)
}

// defaultProbeCandidates returns the baseline chain list the probe
// endpoint enumerates when the caller doesn't pass an explicit
// Networks filter. Order matters only for display; each is probed
// concurrently.
func defaultProbeCandidates() []probeCandidate {
	return []probeCandidate{
		// Bitcoin family. BIP84 native segwit is the modern default
		// where the chain supports it; BCH + DOGE are P2PKH-only.
		// BIP44 legacy address is also surfaced for BTC since many
		// older wallets still use it.
		{
			Network: "bitcoin", Variant: "P2WPKH (BIP84)",
			Curve: "secp256k1", DerivationPath: "m/84'/0'/0'/0/0",
			outputScript: "p2wpkh", addressChain: "bitcoin",
			networkType: "bitcoin", networkChainId: "bitcoin",
		},
		{
			Network: "bitcoin", Variant: "P2PKH (BIP44)",
			Curve: "secp256k1", DerivationPath: "m/44'/0'/0'/0/0",
			outputScript: "p2pkh", addressChain: "bitcoin",
			networkType: "bitcoin", networkChainId: "bitcoin",
		},
		{
			Network: "litecoin", Variant: "P2WPKH (BIP84)",
			Curve: "secp256k1", DerivationPath: "m/84'/2'/0'/0/0",
			outputScript: "p2wpkh", addressChain: "litecoin",
			networkType: "bitcoin", networkChainId: "litecoin",
		},
		{
			Network: "monacoin", Variant: "P2WPKH (BIP84)",
			Curve: "secp256k1", DerivationPath: "m/84'/22'/0'/0/0",
			outputScript: "p2wpkh", addressChain: "monacoin",
			networkType: "bitcoin", networkChainId: "monacoin",
		},
		{
			Network: "monacoin", Variant: "P2PKH (BIP44)",
			Curve: "secp256k1", DerivationPath: "m/44'/22'/0'/0/0",
			outputScript: "p2pkh", addressChain: "monacoin",
			networkType: "bitcoin", networkChainId: "monacoin",
		},
		{
			Network: "bitcoin-cash", Variant: "P2PKH (BIP44)",
			Curve: "secp256k1", DerivationPath: "m/44'/145'/0'/0/0",
			outputScript: "p2pkh", addressChain: "bitcoincash",
			networkType: "bitcoin", networkChainId: "bitcoin-cash",
		},
		{
			Network: "dogecoin", Variant: "P2PKH (BIP44)",
			Curve: "secp256k1", DerivationPath: "m/44'/3'/0'/0/0",
			outputScript: "p2pkh", addressChain: "dogecoin",
			networkType: "bitcoin", networkChainId: "dogecoin",
		},

		// EVM — BIP44 coin_type 60. Same derivation covers every
		// EVM chain (Ethereum, Polygon, Arbitrum, Base, Optimism,
		// BSC, etc.); we probe mainnet Ethereum as the canonical
		// reference. If the user has activity only on an L2,
		// nothing in this probe surfaces it — a follow-up would
		// probe a configured list of EVM chains.
		{
			Network: "ethereum", Variant: "standard",
			Curve: "secp256k1", DerivationPath: "m/44'/60'/0'/0/0",
			outputScript: "eth", addressChain: "",
			networkType: "evm", networkChainId: "1",
		},

		// Solana — both "correct" conventions. Auto-detect picks
		// whichever one has activity; if neither, UI shows both
		// and the user picks.
		{
			Network: "solana", Variant: "sollet (seed[:32])",
			Curve: "ed25519", DerivationPath: "",
			networkType: "solana", networkChainId: "mainnet",
		},
		{
			Network: "solana", Variant: "phantom (m/44'/501'/0'/0')",
			Curve: "ed25519", DerivationPath: "m/44'/501'/0'/0'",
			networkType: "solana", networkChainId: "mainnet",
		},
	}
}

// apiWalletProbeActivity implements Wallet/{id}:probeActivity.
//
//	Keys []*wltsign.KeyDescription  // length 1, decrypts the mnemonic
//	Networks []string               // optional filter — only probe these network tags
//	                                //   ("bitcoin", "ethereum", "solana", …). Empty = all.
//	Timeout  int                    // optional, seconds; default 10
func apiWalletProbeActivity(ctx *apirouter.Context, in struct {
	Keys     []*wltsign.KeyDescription
	Networks []string
	Timeout  int
}) (any, error) {
	e := wltintf.GetEnv(ctx)
	if e == nil {
		return nil, errors.New("failed to get env")
	}
	w := apirouter.GetObject[Wallet](ctx, "Wallet")
	if w == nil {
		return nil, errors.New("Wallet required")
	}
	if len(w.Keys) != 1 || w.Keys[0].Schema != "mnemonic" {
		return nil, errors.New("probeActivity requires a mnemonic-backed wallet")
	}
	if len(in.Keys) != 1 {
		return nil, fmt.Errorf("probeActivity requires exactly 1 KeyDescription, got %d", len(in.Keys))
	}

	share, err := w.Keys[0].decryptMnemonic(in.Keys[0])
	if err != nil {
		return nil, fmt.Errorf("decrypt mnemonic: %w", err)
	}

	mnemonicStr, err := reconstructMnemonic(share)
	if err != nil {
		return nil, err
	}
	seed := bip39.NewSeed(mnemonicStr, share.Passphrase)
	defer zero(seed)

	// Filter candidates by the requested network list (empty = all).
	candidates := defaultProbeCandidates()
	if len(in.Networks) > 0 {
		wanted := map[string]struct{}{}
		for _, n := range in.Networks {
			wanted[strings.ToLower(n)] = struct{}{}
		}
		filtered := candidates[:0]
		for _, c := range candidates {
			if _, ok := wanted[strings.ToLower(c.Network)]; ok {
				filtered = append(filtered, c)
			}
		}
		candidates = filtered
	}

	timeout := time.Duration(in.Timeout) * time.Second
	if timeout <= 0 {
		timeout = 10 * time.Second
	}
	probeCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	results := make([]ProbeResult, len(candidates))
	var wg sync.WaitGroup
	for i, c := range candidates {
		wg.Add(1)
		go func(i int, c probeCandidate) {
			defer wg.Done()
			results[i] = probeOne(probeCtx, seed, c)
		}(i, c)
	}
	wg.Wait()
	return results, nil
}

// probeOne derives a candidate address from the seed and hits its
// chain's RPC for activity. Returns a ProbeResult; any error is
// recorded on Result.Error so the caller gets partial data instead
// of a whole-request failure.
func probeOne(ctx context.Context, seed []byte, c probeCandidate) ProbeResult {
	out := ProbeResult{
		Network:        c.Network,
		Variant:        c.Variant,
		Curve:          c.Curve,
		DerivationPath: c.DerivationPath,
	}

	priv, err := derivePrivkeyFromSeed(seed, c.Curve, c.DerivationPath)
	if err != nil {
		out.Error = fmt.Sprintf("derive: %s", err)
		return out
	}
	defer zero(priv)

	pub, err := derivePub(c.Curve, priv)
	if err != nil {
		out.Error = fmt.Sprintf("derive pub: %s", err)
		return out
	}
	out.Pubkey = base64.RawURLEncoding.EncodeToString(pub)

	addr, err := encodeAddress(c, pub)
	if err != nil {
		out.Error = fmt.Sprintf("encode address: %s", err)
		return out
	}
	out.Address = addr

	bal, hasActivity, err := probeBalance(ctx, c, addr)
	if err != nil {
		out.Error = fmt.Sprintf("probe: %s", err)
		return out
	}
	out.Balance = bal
	out.HasActivity = hasActivity
	return out
}

// encodeAddress formats a derived pubkey as the chain-native address
// string. Uses outscript for Bitcoin-family and EVM; base58 for
// Solana (ed25519 raw pubkey).
func encodeAddress(c probeCandidate, pub []byte) (string, error) {
	switch c.Curve {
	case "ed25519":
		return base58.Bitcoin.Encode(pub), nil
	case "secp256k1":
		pk, err := secp256k1.ParsePubKey(pub)
		if err != nil {
			return "", err
		}
		o, err := outscript.New(pk).Out(c.outputScript)
		if err != nil {
			return "", fmt.Errorf("outscript.Out(%q): %w", c.outputScript, err)
		}
		if c.addressChain == "" {
			// EVM: Address() doesn't take a chain tag
			return o.Address()
		}
		return o.Address(c.addressChain)
	default:
		return "", fmt.Errorf("unsupported curve %q", c.Curve)
	}
}

// probeBalance hits the appropriate RPC method for the candidate's
// chain. Returns the raw balance (in the chain's smallest units as
// a decimal string) and a bool indicating non-zero activity.
func probeBalance(ctx context.Context, c probeCandidate, address string) (balance string, hasActivity bool, err error) {
	// Ephemeral Network — populated with just enough fields that
	// getRPC() wires up the right endpoint (freshness-picked for
	// EVM mainnet, modchain for Bitcoin family, helius for Solana).
	n := &wltnet.Network{
		Type:    c.networkType,
		ChainId: c.networkChainId,
	}

	switch c.networkType {
	case "evm":
		var hexBal string
		raw, err := n.DoRPCCtx(ctx, "eth_getBalance", address, "latest")
		if err != nil {
			return "", false, err
		}
		if err := json.Unmarshal(raw, &hexBal); err != nil {
			return "", false, err
		}
		v, ok := new(big.Int).SetString(hexBal, 0)
		if !ok {
			return "", false, fmt.Errorf("invalid eth_getBalance response %q", hexBal)
		}
		return v.String(), v.Sign() > 0, nil

	case "bitcoin":
		// Bitcoin-family modchain "modchain_assets" returns the
		// asset list for an address. Presence of any item (native
		// coin UTXOs or tokens) indicates activity.
		raw, err := n.DoRPCCtx(ctx, "modchain_assets", address)
		if err != nil {
			return "", false, err
		}
		// The response shape varies; detect activity by "not empty
		// / not null". We surface the raw JSON as balance so the
		// client can render it without another round-trip.
		trimmed := strings.TrimSpace(string(raw))
		empty := trimmed == "" || trimmed == "null" || trimmed == "[]" || trimmed == "{}"
		return trimmed, !empty, nil

	case "solana":
		var resp struct {
			Value uint64 `json:"value"`
		}
		raw, err := n.DoRPCCtx(ctx, "getBalance", address)
		if err != nil {
			return "", false, err
		}
		if err := json.Unmarshal(raw, &resp); err != nil {
			return "", false, err
		}
		return fmt.Sprintf("%d", resp.Value), resp.Value > 0, nil

	default:
		return "", false, fmt.Errorf("unsupported network type %q", c.networkType)
	}
}

// reconstructMnemonic re-encodes stored entropy into the original
// BIP39 mnemonic string. Same logic mnemonicToSeed uses, exposed
// separately so the probe path can keep the raw mnemonic string
// around (bip39.NewSeed wants a string).
func reconstructMnemonic(share *MnemonicKeyShare) (string, error) {
	prevList := bip39.GetWordList()
	if list := wordlistByName(share.Language); list != nil {
		bip39.SetWordList(list)
		defer bip39.SetWordList(prevList)
	}
	m, err := bip39.NewMnemonic(share.Entropy)
	if err != nil {
		return "", fmt.Errorf("bip39: re-encode entropy: %w", err)
	}
	return m, nil
}

File: rust/src/probe.rs
//! `Wallet:probeActivity` core (port of wltwallet/probe.go): derive the
//! canonical addresses for a BIP39 seed across the supported chains and probe
//! each chain's RPC for on-chain activity. Read-only; no wallet state changes.
//!
//! This is the reusable core (seed → candidates → addresses → probe). The FFI
//! endpoint decrypts a mnemonic-keep wallet's share to obtain the seed; that
//! wallet type + handler wiring builds on top of this.

use serde_json::Value;

use crate::{Error, Result};

/// One (chain, BIP variant) address the probe derives + checks.
pub struct Candidate {
    pub network: &'static str,          // "bitcoin" | "ethereum" | "solana" | …
    pub variant: &'static str,          // UI label
    pub curve: &'static str,            // "secp256k1" | "ed25519"
    pub path: &'static str,             // BIP32 path ("" = ed25519 Sollet)
    pub address_chain: &'static str,    // bitcoin::hd_address chain id (secp/bitcoin)
    pub network_type: &'static str,     // "evm" | "bitcoin" | "solana"
    pub network_chain_id: &'static str, // for the probe network
}

/// The baseline candidate list (Go `defaultProbeCandidates`; the P2WPKH/EVM/
/// Solana set — legacy P2PKH variants land with a public P2PKH encoder).
pub fn default_candidates() -> Vec<Candidate> {
    vec![
        Candidate { network: "bitcoin", variant: "P2WPKH (BIP84)", curve: "secp256k1", path: "m/84'/0'/0'/0/0", address_chain: "bitcoin", network_type: "bitcoin", network_chain_id: "bitcoin" },
        Candidate { network: "litecoin", variant: "P2WPKH (BIP84)", curve: "secp256k1", path: "m/84'/2'/0'/0/0", address_chain: "litecoin", network_type: "bitcoin", network_chain_id: "litecoin" },
        Candidate { network: "monacoin", variant: "P2WPKH (BIP84)", curve: "secp256k1", path: "m/84'/22'/0'/0/0", address_chain: "monacoin", network_type: "bitcoin", network_chain_id: "monacoin" },
        Candidate { network: "ethereum", variant: "standard", curve: "secp256k1", path: "m/44'/60'/0'/0/0", address_chain: "", network_type: "evm", network_chain_id: "1" },
        Candidate { network: "solana", variant: "sollet (seed[:32])", curve: "ed25519", path: "", address_chain: "", network_type: "solana", network_chain_id: "mainnet" },
        Candidate { network: "solana", variant: "phantom (m/44'/501'/0'/0')", curve: "ed25519", path: "m/44'/501'/0'/0'", address_chain: "", network_type: "solana", network_chain_id: "mainnet" },
    ]
}

/// Derive `(pubkey_b64url, address)` for a candidate from `seed`.
pub fn derive_address(seed: &[u8], c: &Candidate) -> Result<(String, String)> {
    let pubkey = crate::hdderive::derive_pubkey_for_path(seed, c.curve, c.path)
        .map_err(|e| Error::Env(e.to_string()))?;
    let address = match (c.curve, c.network_type) {
        ("ed25519", _) => bs58::encode(&pubkey).into_string(),
        ("secp256k1", "evm") => crate::hdderive::evm_address(&pubkey).map_err(|e| Error::Env(e.to_string()))?,
        ("secp256k1", "bitcoin") => {
            let p33: [u8; 33] = pubkey.clone().try_into().map_err(|_| Error::Env("secp pubkey not 33 bytes".into()))?;
            crate::bitcoin::hd_address(&p33, c.address_chain)?
        }
        (curve, ty) => return Err(Error::Env(format!("unsupported probe candidate {curve}/{ty}"))),
    };
    use base64::Engine;
    Ok((base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(&pubkey), address))
}

/// Probe a candidate's chain for activity, returning `(raw_balance, has_activity)`.
/// EVM: eth_getBalance (wei). Solana: getBalance (lamports). Bitcoin: presence
/// of any `modchain_assets` entry.
pub fn probe_balance(rpc: &str, c: &Candidate, address: &str) -> Result<(String, bool)> {
    match c.network_type {
        "evm" => {
            let hex = crate::rpc::call(rpc, "eth_getBalance", serde_json::json!([address, "latest"]))?;
            let hex = hex.as_str().ok_or_else(|| Error::Env("eth_getBalance not a string".into()))?;
            let stripped = hex.strip_prefix("0x").unwrap_or(hex);
            let n = num_bigint::BigInt::parse_bytes(stripped.as_bytes(), 16).ok_or_else(|| Error::Env(format!("bad balance hex {hex}")))?;
            Ok((n.to_string(), n.sign() == num_bigint::Sign::Plus))
        }
        "solana" => {
            let res = crate::rpc::call(rpc, "getBalance", serde_json::json!([address]))?;
            let v = res.get("value").and_then(Value::as_u64).unwrap_or(0);
            Ok((v.to_string(), v > 0))
        }
        "bitcoin" => {
            let res = crate::rpc::call(rpc, "modchain_assets", serde_json::json!([address]))?;
            let empty = res.is_null() || res.as_array().map(|a| a.is_empty()).unwrap_or(false) || res.as_object().map(|o| o.is_empty()).unwrap_or(false);
            Ok((res.to_string(), !empty))
        }
        other => Err(Error::Env(format!("unsupported probe network type {other}"))),
    }
}

/// Full probe of one candidate: derive its address then hit its RPC. Errors are
/// captured on the result (partial data), matching Go `probeOne`.
pub fn probe_one(seed: &[u8], c: &Candidate, rpc: &str) -> Value {
    let mut out = serde_json::json!({
        "network": c.network, "variant": c.variant, "curve": c.curve, "derivationPath": c.path,
    });
    let (pubkey, address) = match derive_address(seed, c) {
        Ok(x) => x,
        Err(e) => {
            out["error"] = Value::String(format!("derive: {e}"));
            return out;
        }
    };
    out["pubkey"] = Value::String(pubkey);
    out["address"] = Value::String(address.clone());
    match probe_balance(rpc, c, &address) {
        Ok((balance, activity)) => {
            out["balance"] = Value::String(balance);
            out["hasActivity"] = Value::Bool(activity);
        }
        Err(e) => out["error"] = Value::String(format!("probe: {e}")),
    }
    out
}

/// Filter the default candidates to a requested network list (empty = all),
/// case-insensitive on the `network` tag (Go probe's Networks filter).
pub fn candidates_for(networks: &[String]) -> Vec<Candidate> {
    if networks.is_empty() {
        return default_candidates();
    }
    let wanted: Vec<String> = networks.iter().map(|n| n.to_lowercase()).collect();
    default_candidates().into_iter().filter(|c| wanted.iter().any(|w| w == &c.network.to_lowercase())).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    // The standard "abandon … about" test mnemonic.
    const MNEMONIC: &str = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";

    #[test]
    fn ethereum_candidate_matches_bip44_vector() {
        let seed = crate::bip39::mnemonic_to_seed(MNEMONIC, "");
        let eth = default_candidates().into_iter().find(|c| c.network == "ethereum").unwrap();
        let (_pub, addr) = derive_address(&seed, &eth).unwrap();
        // Well-known BIP44 m/44'/60'/0'/0/0 address for this mnemonic.
        assert_eq!(addr, "0x9858EfFD232B4033E47d90003D41EC34EcaEda94");
    }

    #[test]
    fn solana_and_bitcoin_candidates_derive_addresses() {
        let seed = crate::bip39::mnemonic_to_seed(MNEMONIC, "");
        for c in default_candidates() {
            let (pubkey, addr) = derive_address(&seed, &c).unwrap();
            assert!(!pubkey.is_empty() && !addr.is_empty(), "{}/{}", c.network, c.variant);
            match c.network_type {
                "solana" => assert!(!addr.contains(['0', 'O', 'I', 'l']), "base58 addr {addr}"),
                "bitcoin" if c.address_chain == "bitcoin" => assert!(addr.starts_with("bc1"), "P2WPKH {addr}"),
                "bitcoin" if c.address_chain == "litecoin" => assert!(addr.starts_with("ltc1"), "LTC {addr}"),
                _ => {}
            }
        }
        // The Sollet and Phantom Solana variants differ (different derivation).
        let sollet = default_candidates().into_iter().find(|c| c.variant.starts_with("sollet")).unwrap();
        let phantom = default_candidates().into_iter().find(|c| c.variant.starts_with("phantom")).unwrap();
        assert_ne!(derive_address(&seed, &sollet).unwrap().1, derive_address(&seed, &phantom).unwrap().1);
    }

    #[test]
    fn probe_one_reports_activity() {
        use std::io::{Read, Write};
        use std::net::TcpListener;
        // Mock node reporting a 1 ETH balance.
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let addr = listener.local_addr().unwrap();
        std::thread::spawn(move || {
            if let Ok((mut s, _)) = listener.accept() {
                let mut buf = [0u8; 4096];
                let _ = s.read(&mut buf);
                let body = r#"{"jsonrpc":"2.0","id":1,"result":"0xde0b6b3a7640000"}"#;
                let resp = format!("HTTP/1.1 200 OK\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}", body.len(), body);
                let _ = s.write_all(resp.as_bytes());
            }
        });
        let rpc = format!("http://{addr}");

        let seed = crate::bip39::mnemonic_to_seed(MNEMONIC, "");
        let eth = default_candidates().into_iter().find(|c| c.network == "ethereum").unwrap();
        let res = probe_one(&seed, &eth, &rpc);
        assert_eq!(res["address"], "0x9858EfFD232B4033E47d90003D41EC34EcaEda94");
        assert_eq!(res["balance"], "1000000000000000000");
        assert_eq!(res["hasActivity"], true);
        assert!(res.get("error").is_none());
    }

    #[test]
    fn candidates_filter_by_network() {
        assert_eq!(candidates_for(&["ethereum".into()]).len(), 1);
        assert_eq!(candidates_for(&["solana".into()]).len(), 2);
        assert_eq!(candidates_for(&[]).len(), default_candidates().len());
    }
}
